src/lib/support/BufferReader.h

```diff
@@ -0,0 +1,146 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    Utility classes for safely reading from size-limited buffers.
+ */
+
+#pragma once
+
+#include <climits>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+#include <stdint.h>
+
+namespace chip {
+
+/**
+ *  @class LittleEndianReader
+ *
+ *  Simple reader for reading little-endian things out of buffers.
+ */
+class LittleEndianReader
```

Comment:
I'd recommend following both nlio and CHIPEncoding convention and create both LittleEndian and BigEndian namespaces with a reader class instance in each.

Reply:
Also good catch.  I'm going to do `LittleEndian::Reader` and not worry about the BigEndian version for now.  If we ever need that (which I doubt), we can add it easily.

Reply:
So the issue here is that once I have `chip::LittleEndian::Reader` and `chip::Encoding::LittleEndian::*`, `LittleEndian` becomes ambiguous and more full qualification is needed.  I guess I can do `chip::Encoding::LittleEndian::Reader`... @gerickson any preferences?

Reply:
I like `chip::Encoding::LittleEndian::Reader`.

Reply:
I guess that ambiguity was because I was in code that had `using namespace chip::Encoding`.